src/gsumm.c

```diff
@@ -776,177 +772,56 @@ SEXP gmax(SEXP x, SEXP narm)
-              REAL(sub)[j] = REAL(x)[k];
-            } else {
-              REAL(ans)[i] = NA_REAL;
-              isna = TRUE; break;
-            }
-          } else {
-            u.d = REAL(x)[k];
-            if (u.ll != NA_INT64_LL) {
-              REAL(sub)[j] = (double)u.ll;
-            } else {
-              REAL(ans)[i] = NA_REAL;
-              isna = TRUE; break;
-            }
-          }
-        }
-        if (isna) continue;
-        medianindex = (R_len_t)(ceil((double)(thisgrpsize)/2));
-        REAL(ans)[i] = dquickselect(ptr, thisgrpsize, medianindex-1); // 0-indexed
-        // all elements to the left of thisgrpsize/2 is < the value at that index
-        // we just need to get min of last half
-        if (thisgrpsize % 2 == 0) {
-          val = REAL(sub)[medianindex]; // 0-indexed
-          for (imed=medianindex+1; imed<thisgrpsize; imed++) {
-            val = REAL(sub)[imed] > val ? val : REAL(sub)[imed];
-          }
-          REAL(ans)[i] = (REAL(ans)[i] + val)/2.0;
-        }
-      }
-    } else {
-      for (i=0; i<ngrp; i++) {
-        nacount = 0;
-        thisgrpsize = grpsize[i];
-        for (j=0; j<thisgrpsize; j++) {
-          k = ff[i]+j-1;
-          if (isunsorted) k = oo[k]-1;
-          k = (irowslen == -1) ? k : irows[k]-1;
-          // TODO: raise this if-statement?
-          if (!isint64) {
-            if (!ISNAN(REAL(x)[k])) {
-              REAL(sub)[j-nacount] = REAL(x)[k];
-            } else { nacount++; continue; }
-          } else {
-            u.d = REAL(x)[k];
-            if (u.ll != NA_INT64_LL) {
-              REAL(sub)[j-nacount] = (double)u.ll;
-            } else { nacount++; continue; }
-          }
-        }
-        if (nacount == thisgrpsize) {
-          REAL(ans)[i] = NA_REAL; // all NAs
-          continue;
-        }
-        thisgrpsize -= nacount;
-        SETLENGTH(sub, thisgrpsize);
-        medianindex = (R_len_t)(ceil((double)(thisgrpsize)/2));
-        REAL(ans)[i] = dquickselect(ptr, thisgrpsize, medianindex-1);
-        if (thisgrpsize % 2 == 0) {
-          // all elements to the left of thisgrpsize/2 is < the value at that index
-          // we just need to get min of last half
-          val = REAL(sub)[medianindex]; // 0-indexed
-          for (imed=medianindex+1; imed<thisgrpsize; imed++) {
-            val = REAL(sub)[imed] > val ? val : REAL(sub)[imed];
-          }
-          REAL(ans)[i] = (REAL(ans)[i] + val)/2.0;
-        }
+  case REALSXP: {
+    double *subd = REAL(PROTECT(allocVector(REALSXP, maxgrpn))); // allocate once upfront and reuse
+    int64_t *xi64 = (int64_t *)REAL(x);
+    double  *xd = REAL(x);
+    for (int i=0; i<ngrp; ++i) {
+      int thisgrpsize = grpsize[i], nacount=0;
+      for (int j=0; j<thisgrpsize; ++j) {
+        int k = ff[i]+j-1;
+        if (isunsorted) k = oo[k]-1;
+        k = (irowslen == -1) ? k : irows[k]-1;
+        if (isInt64 ? xi64[k]==NA_INTEGER64 : ISNAN(xd[k])) nacount++;
```

Comment:
Isn't that better to move int64 check outside of the loop and duplicate loop to avoid extra branching in each iteration?

Reply:
Further comment here: https://github.com/Rdatatable/data.table/pull/3564#issuecomment-492925130.  Added here otherwise it doesn't look like I replied when viewing the diff.